crates/client/rpc/src/errors.rs

```diff
@@ -1,67 +1,139 @@
-    fn from(err: StarknetTransactionExecutionError) -> Self {
-        match err {
-            StarknetTransactionExecutionError::ContractNotFound => StarknetRpcApiError::ContractNotFound,
-            StarknetTransactionExecutionError::ClassAlreadyDeclared => StarknetRpcApiError::ClassAlreadyDeclared,
-            StarknetTransactionExecutionError::ClassHashNotFound => StarknetRpcApiError::ClassHashNotFound,
-            StarknetTransactionExecutionError::InvalidContractClass => StarknetRpcApiError::InvalidContractClass,
-            StarknetTransactionExecutionError::ContractError => StarknetRpcApiError::ContractError,
-        }
+#[derive(Debug, Error)]
+#[error("Contract Error")]
+pub struct ContractError {
+    revert_error: String,
+}
+
+#[derive(Debug)]
+pub struct DisplayableContractAddress(pub ContractAddress);
+
+impl fmt::Display for DisplayableContractAddress {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // unwrapping all the layers of different types to get display from starkhash
+        let address = &self.0.0.0;
+        write!(f, "{}", address)
+    }
+}
+
+impl From<ContractAddress> for DisplayableContractAddress {
+    fn from(value: ContractAddress) -> Self {
+        DisplayableContractAddress(value)
     }
 }
 
 impl From<StarknetRpcApiError> for jsonrpsee::core::Error {
     fn from(err: StarknetRpcApiError) -> Self {
-        // TODO:
-        // when able to ge rich errors out of runtime,
-        // match and populate the `data` field
-        jsonrpsee::core::Error::Call(CallError::Custom(ErrorObject::owned(err as i32, err.to_string(), None::<()>)))
+        let code = match err {
+            StarknetRpcApiError::FailedToReceiveTxn => 1,
+            StarknetRpcApiError::ContractNotFound(_) => 20,
+            StarknetRpcApiError::BlockNotFound => 24,
+            StarknetRpcApiError::InvalidTxnIndex => 27,
+            StarknetRpcApiError::ClassHashNotFound => 28,
+            StarknetRpcApiError::TxnHashNotFound => 29,
+            StarknetRpcApiError::PageSizeTooBig => 31,
+            StarknetRpcApiError::NoBlocks => 32,
+            StarknetRpcApiError::InvalidContinuationToken => 33,
+            StarknetRpcApiError::TooManyKeysInFilter => 34,
+            StarknetRpcApiError::FailedToFetchPendingTransactions => 38,
+            StarknetRpcApiError::ContractError(_) => 40,
+            StarknetRpcApiError::InvalidContractClass => 50,
+            StarknetRpcApiError::ClassAlreadyDeclared => 51,
+            StarknetRpcApiError::ValidationFailure => 55,
+            StarknetRpcApiError::UnsupportedTxVersion => 61,
+            StarknetRpcApiError::InternalServerError => 500,
+            StarknetRpcApiError::UnimplementedMethod => 501,
+            StarknetRpcApiError::ProofLimitExceeded => 10000,
+        };
+
+        let data = match &err {
+            StarknetRpcApiError::ContractError(ref error) => Some(serde_json::json!({
```

Comment:
should we also implement Display for `ContractError` so we don't need to match here?

Reply:
It's not really a display

Reply:
https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_api_openrpc.json#L3948-L3963

The specs specify a different field, while the message stays `ContractError`

Reply:
@apoorvsadana yep, the `data` field contains a json with a `revert_error` field